django/db/models/fields/related.py

```diff
@@ -1041,37 +1049,119 @@ def __class_getitem__(cls, *args, **kwargs):
+                        "no db_default value.",
+                        hint="Set a db_default value, or change the on_delete rule.",
+                        obj=self,
+                        id="fields.E322",
+                    )
+                )
+            for db in databases:
+                if not router.allow_migrate_model(db, self.model):
+                    continue
+                connection = connections[db]
+                if not (
+                    "supports_on_delete_db_default"
+                    in self.model._meta.required_db_features
+                    or connection.features.supports_on_delete_db_default
+                ):
+                    errors.append(
+                        checks.Error(
+                            f"{connection.display_name} does not support a "
+                            "DB_SET_DEFAULT.",
+                            hint="Change the on_delete rule to SET_DEFAULT.",
+                            obj=self,
+                            id="fields.E324",
+                        ),
+                    )
+        elif on_delete == DB_RESTRICT:
+            for db in databases:
+                if not router.allow_migrate_model(db, self.model):
+                    continue
+                connection = connections[db]
+                if not (
+                    "supports_on_delete_db_restrict"
+                    in self.model._meta.required_db_features
+                    or connection.features.supports_on_delete_db_restrict
+                ):
+                    errors.append(
+                        checks.Error(
+                            f"{connection.display_name} does not support a "
+                            "DB_RESTRICT.",
+                            hint="Change the on_delete rule to RESTRICT.",
+                            obj=self,
+                            id="fields.E324",
+                        ),
+                    )
+        elif not isinstance(self.remote_field.model, str):
+            # Database and Python variants cannot be mixed in a chain of
+            # model references.
+            is_db_on_delete = isinstance(on_delete, DatabaseOnDelete)
+            ref_model_related_fields = (
+                ref_model_field.remote_field
+                for ref_model_field in self.remote_field.model._meta.get_fields()
+                if ref_model_field.related_model and not ref_model_field.auto_created
+            )
+
+            for ref_remote_field in ref_model_related_fields:
+                if (
+                    ref_remote_field.on_delete is not None
+                    and isinstance(ref_remote_field.on_delete, DatabaseOnDelete)
+                    is not is_db_on_delete
+                ):
+                    on_delete_type = "database" if is_db_on_delete else "python"
+                    ref_on_delete_type = "python" if is_db_on_delete else "database"
+                    errors.append(
+                        checks.Error(
```

Comment:
Is there any harm in letting DO_NOTHING pass through here when there are DB_* variants? (Unless we want to create a DB_DO_NOTHING?)

Reply:
Agreed. Updated with extra tests.